File: diffmerge-2/py/diffmerge/diff.py
"""Diff algorithms for diffmerge library."""

import re
from typing import List, Optional, Tuple
from .types import DiffHunk, DiffResult, DiffStats, DiffOptions
from .utils import split_lines


def _lcs(a: List[str], b: List[str]) -> List[Tuple[int, int]]:
    """Compute Longest Common Subsequence using dynamic programming.

    Returns list of (index_in_a, index_in_b) pairs for matching elements.
    """
    m, n = len(a), len(b)

    if m == 0 or n == 0:
        return []

    # Build LCS length table
    dp = [[0] * (n + 1) for _ in range(m + 1)]

    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if a[i - 1] == b[j - 1]:
                dp[i][j] = dp[i - 1][j - 1] + 1
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])

    # Backtrack to find LCS
    result = []
    i, j = m, n
    while i > 0 and j > 0:
        if a[i - 1] == b[j - 1]:
            result.append((i - 1, j - 1))
            i -= 1
            j -= 1
        elif dp[i - 1][j] > dp[i][j - 1]:
            i -= 1
        else:
            j -= 1

    result.reverse()
    return result


def _normalize_line(line: str, ignore_whitespace: bool) -> str:
    """Normalize a line for comparison."""
    if ignore_whitespace:
        return line.strip()
    return line


def _lines_equal(a: str, b: str, ignore_whitespace: bool) -> bool:
    """Check if two lines are equal, optionally ignoring whitespace."""
    if ignore_whitespace:
        return a.strip() == b.strip()
    return a == b


def diff_lines(old: str, new: str, options: Optional[DiffOptions] = None) -> DiffResult:
    """Compute line-by-line diff using LCS algorithm.

    Args:
        old: Original content
        new: New content
        options: Diff options (ignore_whitespace, ignore_blank_lines, context_lines)

    Returns:
        DiffResult with hunks and stats
    """
    opts = options or {}
    ignore_whitespace = opts.get("ignore_whitespace", False)
    ignore_blank_lines = opts.get("ignore_blank_lines", False)
    context_lines = opts.get("context_lines", 3)

    old_lines = split_lines(old)
    new_lines = split_lines(new)

    # For comparison, optionally normalize lines
    def should_include(line: str) -> bool:
        if ignore_blank_lines and line.strip() == "":
            return False
        return True

    # Create filtered versions for LCS comparison
    if ignore_blank_lines:
        old_filtered = [(i, line) for i, line in enumerate(old_lines) if should_include(line)]
        new_filtered = [(i, line) for i, line in enumerate(new_lines) if should_include(line)]
    else:
        old_filtered = [(i, line) for i, line in enumerate(old_lines)]
        new_filtered = [(i, line) for i, line in enumerate(new_lines)]

    # Normalize for comparison
    old_compare = [_normalize_line(line, ignore_whitespace) for _, line in old_filtered]
    new_compare = [_normalize_line(line, ignore_whitespace) for _, line in new_filtered]

    # Compute LCS on filtered/normalized lines
    lcs = _lcs(old_compare, new_compare)

    # Convert LCS indices back to original indices
    lcs_original = []
    for old_idx, new_idx in lcs:
        orig_old_idx = old_filtered[old_idx][0]
        orig_new_idx = new_filtered[new_idx][0]
        lcs_original.append((orig_old_idx, orig_new_idx))

    # Build diff hunks from LCS
    hunks: List[DiffHunk] = []
    old_pos = 0
    new_pos = 0

    for old_idx, new_idx in lcs_original:
        # Add deletions (lines in old not in LCS)
        while old_pos < old_idx:
            hunks.append({
                "op": "delete",
                "content": old_lines[old_pos],
                "old_start": old_pos + 1,
                "old_count": 1
            })
            old_pos += 1

        # Add insertions (lines in new not in LCS)
        while new_pos < new_idx:
            hunks.append({
                "op": "insert",
                "content": new_lines[new_pos],
                "new_start": new_pos + 1,
                "new_count": 1
            })
            new_pos += 1

        # Add equal line
        hunks.append({
            "op": "equal",
            "content": new_lines[new_pos],
            "old_start": old_pos + 1,
            "new_start": new_pos + 1,
            "old_count": 1,
            "new_count": 1
        })
        old_pos += 1
        new_pos += 1

    # Handle remaining lines after last LCS match
    while old_pos < len(old_lines):
        hunks.append({
            "op": "delete",
            "content": old_lines[old_pos],
            "old_start": old_pos + 1,
            "old_count": 1
        })
        old_pos += 1

    while new_pos < len(new_lines):
        hunks.append({
            "op": "insert",
            "content": new_lines[new_pos],
            "new_start": new_pos + 1,
            "new_count": 1
        })
        new_pos += 1

    # Apply context_lines filtering if specified
    if context_lines >= 0 and hunks:
        hunks = _filter_context(hunks, context_lines)

    # Calculate stats (excluding blank lines if ignore_blank_lines is True)
    def should_count(h: DiffHunk) -> bool:
        if ignore_blank_lines and h["content"].strip() == "":
            return False
        return True

    additions = sum(1 for h in hunks if h["op"] == "insert" and should_count(h))
    deletions = sum(1 for h in hunks if h["op"] == "delete" and should_count(h))

    stats: DiffStats = {
        "additions": additions,
        "deletions": deletions,
        "changes": min(additions, deletions)
    }

    return {"hunks": hunks, "stats": stats}


def _filter_context(hunks: List[DiffHunk], context_lines: int) -> List[DiffHunk]:
    """Filter hunks to only include equal lines within context_lines of changes."""
    if not hunks:
        return hunks

    # Mark indices of change hunks
    change_indices = set()
    for i, h in enumerate(hunks):
        if h["op"] != "equal":
            change_indices.add(i)

    # Mark equal lines within context_lines of changes
    include_indices = set()
    for i in change_indices:
        include_indices.add(i)
        # Add context before
        for j in range(max(0, i - context_lines), i):
            include_indices.add(j)
        # Add context after
        for j in range(i + 1, min(len(hunks), i + context_lines + 1)):
            include_indices.add(j)

    return [hunks[i] for i in sorted(include_indices)]


def diff_words(old: str, new: str) -> List[DiffHunk]:
    """Compute word-by-word diff.

    Args:
        old: Original text
        new: New text

    Returns:
        List of DiffHunks
    """
    # Split into tokens (words and punctuation/whitespace)
    def tokenize(text: str) -> List[str]:
        # Split on word boundaries, keeping delimiters
        tokens = re.split(r'(\s+|\b)', text)
        return [t for t in tokens if t]

    old_tokens = tokenize(old)
    new_tokens = tokenize(new)

    lcs = _lcs(old_tokens, new_tokens)

    hunks: List[DiffHunk] = []
    old_pos = 0
    new_pos = 0

    for old_idx, new_idx in lcs:
        # Deletions
        if old_pos < old_idx:
            deleted = "".join(old_tokens[old_pos:old_idx])
            hunks.append({"op": "delete", "content": deleted})
            old_pos = old_idx

        # Insertions
        if new_pos < new_idx:
            inserted = "".join(new_tokens[new_pos:new_idx])
            hunks.append({"op": "insert", "content": inserted})
            new_pos = new_idx

        # Equal
        hunks.append({"op": "equal", "content": old_tokens[old_pos]})
        old_pos += 1
        new_pos += 1

    # Remaining deletions
    if old_pos < len(old_tokens):
        deleted = "".join(old_tokens[old_pos:])
        hunks.append({"op": "delete", "content": deleted})

    # Remaining insertions
    if new_pos < len(new_tokens):
        inserted = "".join(new_tokens[new_pos:])
        hunks.append({"op": "insert", "content": inserted})

    return hunks


def diff_chars(old: str, new: str) -> List[DiffHunk]:
    """Compute character-by-character diff.

    Args:
        old: Original text
        new: New text

    Returns:
        List of DiffHunks
    """
    old_chars = list(old)
    new_chars = list(new)

    lcs = _lcs(old_chars, new_chars)

    hunks: List[DiffHunk] = []
    old_pos = 0
    new_pos = 0

    for old_idx, new_idx in lcs:
        # Deletions
        if old_pos < old_idx:
            deleted = "".join(old_chars[old_pos:old_idx])
            hunks.append({"op": "delete", "content": deleted})
            old_pos = old_idx

        # Insertions
        if new_pos < new_idx:
            inserted = "".join(new_chars[new_pos:new_idx])
            hunks.append({"op": "insert", "content": inserted})
            new_pos = new_idx

        # Equal - accumulate consecutive equal chars
        if hunks and hunks[-1]["op"] == "equal":
            hunks[-1]["content"] += old_chars[old_pos]
        else:
            hunks.append({"op": "equal", "content": old_chars[old_pos]})
        old_pos += 1
        new_pos += 1

    # Remaining deletions
    if old_pos < len(old_chars):
        deleted = "".join(old_chars[old_pos:])
        hunks.append({"op": "delete", "content": deleted})

    # Remaining insertions
    if new_pos < len(new_chars):
        inserted = "".join(new_chars[new_pos:])
        hunks.append({"op": "insert", "content": inserted})

    return hunks

File: diffmerge-2/go/dm/diff.go
package dm

import (
	"strings"
	"unicode"
)

// lcs computes the Longest Common Subsequence of two slices
func lcs(a, b []string) []string {
	m := len(a)
	n := len(b)

	// Create DP table
	dp := make([][]int, m+1)
	for i := range dp {
		dp[i] = make([]int, n+1)
	}

	// Fill DP table
	for i := 1; i <= m; i++ {
		for j := 1; j <= n; j++ {
			if a[i-1] == b[j-1] {
				dp[i][j] = dp[i-1][j-1] + 1
			} else {
				dp[i][j] = max(dp[i-1][j], dp[i][j-1])
			}
		}
	}

	// Backtrack to find LCS
	result := make([]string, 0, dp[m][n])
	i, j := m, n
	for i > 0 && j > 0 {
		if a[i-1] == b[j-1] {
			result = append([]string{a[i-1]}, result...)
			i--
			j--
		} else if dp[i-1][j] > dp[i][j-1] {
			i--
		} else {
			j--
		}
	}

	return result
}

// DiffLines computes a line-by-line diff
func DiffLines(old, new string, options DiffOptions) DiffResult {
	// Handle context lines default
	if options.ContextLines == 0 {
		options.ContextLines = 3
	}

	oldLines := SplitLines(old)
	newLines := SplitLines(new)

	// Create comparison versions for ignore options
	oldCompare := make([]string, len(oldLines))
	newCompare := make([]string, len(newLines))

	for i, line := range oldLines {
		if options.IgnoreWhitespace {
			oldCompare[i] = strings.TrimSpace(line)
		} else {
			oldCompare[i] = line
		}
	}

	for i, line := range newLines {
		if options.IgnoreWhitespace {
			newCompare[i] = strings.TrimSpace(line)
		} else {
			newCompare[i] = line
		}
	}

	// Filter blank lines if needed
	var oldFiltered, newFiltered []int
	var oldCompareFiltered, newCompareFiltered []string

	if options.IgnoreBlankLines {
		for i, line := range oldCompare {
			if strings.TrimSpace(line) != "" {
				oldFiltered = append(oldFiltered, i)
				oldCompareFiltered = append(oldCompareFiltered, line)
			}
		}
		for i, line := range newCompare {
			if strings.TrimSpace(line) != "" {
				newFiltered = append(newFiltered, i)
				newCompareFiltered = append(newCompareFiltered, line)
			}
		}
	} else {
		for i := range oldCompare {
			oldFiltered = append(oldFiltered, i)
		}
		oldCompareFiltered = oldCompare
		for i := range newCompare {
			newFiltered = append(newFiltered, i)
		}
		newCompareFiltered = newCompare
	}

	// Compute LCS on the filtered/compared versions
	common := lcs(oldCompareFiltered, newCompareFiltered)

	// Build hunks by tracking positions
	var hunks []DiffHunk
	additions := 0
	deletions := 0

	oldIdx := 0
	newIdx := 0
	commonIdx := 0

	for oldIdx < len(oldFiltered) || newIdx < len(newFiltered) {
		if commonIdx < len(common) {
			// Find next common line in old
			oldNext := -1
			for i := oldIdx; i < len(oldFiltered); i++ {
				if oldCompareFiltered[i] == common[commonIdx] {
					oldNext = i
					break
				}
			}

			// Find next common line in new
			newNext := -1
			for i := newIdx; i < len(newFiltered); i++ {
				if newCompareFiltered[i] == common[commonIdx] {
					newNext = i
					break
				}
			}

			// Add deletions before the common line
			for oldIdx < oldNext {
				realIdx := oldFiltered[oldIdx]
				hunks = append(hunks, DiffHunk{
					Op:       OpDelete,
					Content:  oldLines[realIdx],
					OldStart: realIdx + 1,
					OldCount: 1,
				})
				deletions++
				oldIdx++
			}

			// Add insertions before the common line
			for newIdx < newNext {
				realIdx := newFiltered[newIdx]
				hunks = append(hunks, DiffHunk{
					Op:       OpInsert,
					Content:  newLines[realIdx],
					NewStart: realIdx + 1,
					NewCount: 1,
				})
				additions++
				newIdx++
			}

			// Add the equal line
			oldRealIdx := oldFiltered[oldIdx]
			newRealIdx := newFiltered[newIdx]
			hunks = append(hunks, DiffHunk{
				Op:       OpEqual,
				Content:  oldLines[oldRealIdx],
				OldStart: oldRealIdx + 1,
				NewStart: newRealIdx + 1,
				OldCount: 1,
				NewCount: 1,
			})
			oldIdx++
			newIdx++
			commonIdx++
		} else {
			// No more common lines - rest are deletions/insertions
			for oldIdx < len(oldFiltered) {
				realIdx := oldFiltered[oldIdx]
				hunks = append(hunks, DiffHunk{
					Op:       OpDelete,
					Content:  oldLines[realIdx],
					OldStart: realIdx + 1,
					OldCount: 1,
				})
				deletions++
				oldIdx++
			}
			for newIdx < len(newFiltered) {
				realIdx := newFiltered[newIdx]
				hunks = append(hunks, DiffHunk{
					Op:       OpInsert,
					Content:  newLines[realIdx],
					NewStart: realIdx + 1,
					NewCount: 1,
				})
				additions++
				newIdx++
			}
		}
	}

	// Filter hunks based on context if needed
	if options.ContextLines > 0 && len(hunks) > 0 {
		hunks = filterHunksByContext(hunks, options.ContextLines)
	}

	return DiffResult{
		Hunks: hunks,
		Stats: DiffStats{
			Additions: additions,
			Deletions: deletions,
			Changes:   min(additions, deletions),
		},
	}
}

// filterHunksByContext keeps only hunks within context of changes
func filterHunksByContext(hunks []DiffHunk, contextLines int) []DiffHunk {
	if len(hunks) == 0 {
		return hunks
	}

	// Find indices of change hunks
	var changeIndices []int
	for i, h := range hunks {
		if h.Op != OpEqual {
			changeIndices = append(changeIndices, i)
		}
	}

	if len(changeIndices) == 0 {
		return hunks
	}

	// Mark which hunks to keep
	keep := make([]bool, len(hunks))
	for _, ci := range changeIndices {
		keep[ci] = true
		// Keep context before
		for i := ci - 1; i >= 0 && i >= ci-contextLines; i-- {
			keep[i] = true
		}
		// Keep context after
		for i := ci + 1; i < len(hunks) && i <= ci+contextLines; i++ {
			keep[i] = true
		}
	}

	var result []DiffHunk
	for i, h := range hunks {
		if keep[i] {
			result = append(result, h)
		}
	}

	return result
}

// tokenizeWords splits text into word tokens
func tokenizeWords(text string) []string {
	var tokens []string
	var current strings.Builder

	for _, r := range text {
		if unicode.IsSpace(r) {
			if current.Len() > 0 {
				tokens = append(tokens, current.String())
				current.Reset()
			}
			tokens = append(tokens, string(r))
		} else if unicode.IsPunct(r) {
			if current.Len() > 0 {
				tokens = append(tokens, current.String())
				current.Reset()
			}
			tokens = append(tokens, string(r))
		} else {
			current.WriteRune(r)
		}
	}

	if current.Len() > 0 {
		tokens = append(tokens, current.String())
	}

	return tokens
}

// DiffWords computes a word-by-word diff
func DiffWords(old, new string) []DiffHunk {
	oldTokens := tokenizeWords(old)
	newTokens := tokenizeWords(new)

	return diffTokens(oldTokens, newTokens)
}

// DiffChars computes a character-by-character diff
func DiffChars(old, new string) []DiffHunk {
	oldChars := []rune(old)
	newChars := []rune(new)

	oldStrs := make([]string, len(oldChars))
	newStrs := make([]string, len(newChars))

	for i, r := range oldChars {
		oldStrs[i] = string(r)
	}
	for i, r := range newChars {
		newStrs[i] = string(r)
	}

	return diffTokens(oldStrs, newStrs)
}

// diffTokens computes a diff between token sequences
func diffTokens(oldTokens, newTokens []string) []DiffHunk {
	common := lcs(oldTokens, newTokens)

	var hunks []DiffHunk

	oldIdx := 0
	newIdx := 0
	commonIdx := 0

	for oldIdx < len(oldTokens) || newIdx < len(newTokens) {
		if commonIdx < len(common) {
			// Find next common token in old
			oldNext := -1
			for i := oldIdx; i < len(oldTokens); i++ {
				if oldTokens[i] == common[commonIdx] {
					oldNext = i
					break
				}
			}

			// Find next common token in new
			newNext := -1
			for i := newIdx; i < len(newTokens); i++ {
				if newTokens[i] == common[commonIdx] {
					newNext = i
					break
				}
			}

			// Add deletions before
			if oldNext > oldIdx {
				content := strings.Join(oldTokens[oldIdx:oldNext], "")
				hunks = append(hunks, DiffHunk{
					Op:      OpDelete,
					Content: content,
				})
			}

			// Add insertions before
			if newNext > newIdx {
				content := strings.Join(newTokens[newIdx:newNext], "")
				hunks = append(hunks, DiffHunk{
					Op:      OpInsert,
					Content: content,
				})
			}

			// Add equal token(s) - merge consecutive equals
			equalStart := commonIdx
			equalOldStart := oldNext
			equalNewStart := newNext

			for commonIdx < len(common) {
				nextOld := -1
				for i := equalOldStart + (commonIdx - equalStart); i < len(oldTokens); i++ {
					if oldTokens[i] == common[commonIdx] {
						nextOld = i
						break
					}
				}

				nextNew := -1
				for i := equalNewStart + (commonIdx - equalStart); i < len(newTokens); i++ {
					if newTokens[i] == common[commonIdx] {
						nextNew = i
						break
					}
				}

				if nextOld == equalOldStart+(commonIdx-equalStart) && nextNew == equalNewStart+(commonIdx-equalStart) {
					commonIdx++
				} else {
					break
				}
			}

			content := strings.Join(common[equalStart:commonIdx], "")
			hunks = append(hunks, DiffHunk{
				Op:      OpEqual,
				Content: content,
			})

			oldIdx = oldNext + (commonIdx - equalStart)
			newIdx = newNext + (commonIdx - equalStart)
		} else {
			// No more common tokens
			if oldIdx < len(oldTokens) {
				content := strings.Join(oldTokens[oldIdx:], "")
				hunks = append(hunks, DiffHunk{
					Op:      OpDelete,
					Content: content,
				})
				oldIdx = len(oldTokens)
			}
			if newIdx < len(newTokens) {
				content := strings.Join(newTokens[newIdx:], "")
				hunks = append(hunks, DiffHunk{
					Op:      OpInsert,
					Content: content,
				})
				newIdx = len(newTokens)
			}
		}
	}

	// Merge consecutive hunks of same type
	if len(hunks) > 0 {
		var merged []DiffHunk
		current := hunks[0]

		for i := 1; i < len(hunks); i++ {
			if hunks[i].Op == current.Op {
				current.Content += hunks[i].Content
			} else {
				merged = append(merged, current)
				current = hunks[i]
			}
		}
		merged = append(merged, current)
		hunks = merged
	}

	return hunks
}
